src/rust/engine/fs/src/directory.rs

```diff
@@ -539,19 +617,72 @@ impl DigestTrie {
       ));
       f(&PathBuf::new(), &root);
     }
-    self.walk_helper(PathBuf::new(), f)
+    self.walk_helper(self, PathBuf::new(), symlink_behavior, 0, f)
   }
 
-  fn walk_helper(&self, path_so_far: PathBuf, f: &mut impl FnMut(&Path, &Entry)) {
+  fn walk_helper(
+    &self,
+    root: &DigestTrie,
+    path_so_far: PathBuf,
+    symlink_behavior: SymlinkBehavior,
+    link_depth: LinkDepth,
+    f: &mut impl FnMut(&Path, &Entry),
+  ) {
     for entry in &*self.0 {
-      let path = path_so_far.join(entry.name().as_ref());
-      f(&path, entry);
-      if let Entry::Directory(d) = entry {
-        d.tree.walk_helper(path, f);
-      }
+      self.walk_entry(
+        root,
+        entry,
+        &path_so_far.join(entry.name().as_ref()),
+        symlink_behavior,
+        link_depth,
+        f,
+      );
     }
   }
 
+  fn walk_entry(
+    &self,
+    root: &DigestTrie,
+    entry: &Entry,
+    path: &PathBuf,
+    symlink_behavior: SymlinkBehavior,
+    link_depth: LinkDepth,
+    f: &mut impl FnMut(&Path, &Entry),
+  ) {
+    match (&symlink_behavior, entry) {
+      (SymlinkBehavior::Oblivious, Entry::Symlink(s)) => {
+        let mut tree = self;
+        let mut target = s.target.clone();
+        if target.starts_with("../") {
+          tree = root;
+          target = path.parent().unwrap().join(s.target.clone());
+        }
+
+        if let Ok(Some(target_entry)) = tree.entry_helper(root, &target, link_depth) {
+          if link_depth >= MAX_LINK_DEPTH {
+            // We don't return a Result, so log and move on
+            warn!("Exceeded the maximum link depth while traversing links. Stopping traversal.");
```

Comment:
This should probably include some information on the path that was traversed?

Reply:
I'm guessing we'd want to trim the path. Suggestion on some Rust-y code that'd do that?